isy-serviceapi-core/src/main/java/de/bund/bva/isyfact/serviceapi/core/aop/StelltLoggingKontextBereitInterceptor.java

```diff
@@ -84,6 +84,15 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
         try {
             LOG.debug("Setze Korrelations-ID: " + korrelationsId);
             MdcHelper.pushKorrelationsId(korrelationsId);
+            if (stelltLoggingKontextBereit == null || stelltLoggingKontextBereit.nutzeAufrufKontext()) {
+                Optional<AufrufKontextTo> aufrufKontextToOptional =
+                        aufrufKontextToResolver.leseAufrufKontextTo(invocation.getArguments());
+                aufrufKontextToOptional.ifPresent(aufrufKontextTo -> {
+                    if (aufrufKontextTo.getKorrelationsId() == null || aufrufKontextTo.getKorrelationsId().isEmpty()) {
+                        aufrufKontextTo.setKorrelationsId(korrelationsId);
```

Comment:
if the `korrelationsId` is set in the `aufrufKontextTo` isn't it not necessary to remove it in the `finally-block` like it is done with the MDC in line 99 too?

Reply:
In contrast to the AufrufKontextTo the MDC/MdcHelper is used in a static and therefore more global way, making it necessary to remove the korrelationsId after the invocation. The AufrufKontextTo is only used directly within and for a request and will therefore be cleared after, making it unnecessary to clear the korrelationsId inside the finally block after the method invocation. This is the same behaviour as implemented here:
https://github.com/IsyFact/isyfact-standards/blob/72b53145878c4223a2482d729e3c9e9eed55eebb/isy-serviceapi-core/src/main/java/de/bund/bva/pliscommon/serviceapi/core/aop/StelltLoggingKontextBereitInterceptor.java#L83-L110
(isyfact-standards 1.11.0)

Reply:
I will agree with the argument the MDC is used in a static and more global way than the local `aufrufKontextTo`. 

But the reason for this ticket is the missing `korrelationId` in the `AurufKontextTo` read from the `AufrufKontextResolver`, what in consequence means the content of the `AufrufKontextTo` is read outside of this method context too.

To keep the content in the MDC and in the `AufrufKontextTo` consistent on explicit manner even after their usage,, it will be good practice to remove the `korrelationId` in both container instances. This will reduce an accidental and possible inconsistent use in case of future changes.

Reply:
Der Grund, warum wir das AufrufKontextTo nicht bereinigen, besteht aus den folgenden Punkten:
 
Das AufrufKontextTo ist nur als Transportobjekt gedacht und steht nur für die Laufzeit des Requests zur Verfügung, wird also zum Ende des Requests von der JVM abgeräumt. Der Zweck des AufrufKontextTo besteht nicht darin, weiter im Anwendungskern verwendet zu werden. Das AufrufKontextTo wird nur an der Exception- bzw. Service-Fassade verwendet, um letztendlich die entsprechenden Objekte, z.B. den AufrufKontext im AufrufKontextVerwalter, zu befüllen, die dann weiter im Anwendungskern genutzt werden. Im Anwendungskern wird also entsprechend das durch das AufrufKontextTo befüllte AufrufKontext-Objekt und nicht das AufrufKontextTo-Objekt direkt verwendet.
Dementsprechend entfällt die Notwendigkeit, das AufrufKontextTo-Objekt manuell vor Ende des Requests zu bereinigen.
 
Eine Implementierung wäre zwar technisch möglich, würde allerdings mit einer unnötigen Komplexität einhergehen, die die potenzielle Fehleranfälligkeit erhöhen und die Wartung erschweren könnte.

Reply:
Aber liegt nicht genau in Teilen dieser Argumentation die Begründung das `AufrufKontextTo` zu bereinigen?

_...Das AufrufKontextTo wird nur an der Exception- bzw. Service-Fassade verwendet, um letztendlich die entsprechenden Objekte, z.B. den AufrufKontext im AufrufKontextVerwalter, zu befüllen, die dann weiter im Anwendungskern genutzt werden..._

Wird nun in der Exception- bzw Service-Fassade ein Logoutput erzeugt, dann muss die KorrelationId stimmen, sonst sind Nachverfolgungen von welchem System der Output erzeugt wurde nicht möglich. Und wenn dann über das TransportObjekt die entsprechenden Objekte im AufrufKontextVerwalter, der übrigens analog dem MDC übergreifend genutzt wird, auch noch initialisiert werden, dann ist es um so wichtiger, dass die KorrelationID auch wieder zurückgesetzt bzw. korrigiert wird.
Man stelle sich einfach nur innerhalb eines Requests eine Aufrufkette verschiedener Services im gleichen Anwendungskern vor. Da muss jeder Aufruf über die Servicegrenzen hinweg mit einer eindeutigen KorrelationID versehen und im Log nachvollziehbar gehalten werden. Dazu gehört auch beim Beenden eines Serviceaufrufes, bei dem zuvor die KorrelationID ergänzt wurde, diese Ergänzung nun wieder zurückzunehmen und in den betroffenen Objekten insbesondere im AufrufKontextVerwalter zu korrigieren und das wird nun mal über das TransportObjekt gemacht, oder nicht?

Reply:
siehe Kommentar im Ticket zur Lösung der Problematik